lib/Sema/CSDiagnostics.cpp

```diff
@@ -96,7 +97,20 @@ Type FailureDiagnostic::getRawType(ASTNode node) const {
 
 Type FailureDiagnostic::resolveType(Type rawType, bool reconstituteSugar,
                                     bool wantRValue) const {
-  rawType = rawType.transformRec([&](Type type) -> std::optional<Type> {
+  class Transform : public TypeTransform<Transform> {
+    llvm::function_ref<std::optional<Type>(TypeBase *)> fn;
+
+  public:
+    explicit Transform(llvm::function_ref<std::optional<Type>(TypeBase *)> fn,
+                       ASTContext &ctx)
+        : TypeTransform(ctx), fn(fn) {}
+    std::optional<Type> transform(TypeBase *type, TypePosition position) {
+      return fn(type);
```

Comment:
You don't need to pass in `fn` at this point; just stick it's body inside your Transform.